[SPARK-52619][SQL] Cast TimeType to IntegralType

fartzy opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

 
  Implementation of [SPARK-5261](https://issues.apache.org/jira/browse/SPARK-52619) which is for casting from `TimeType ` i.e. `TIME` data type to `IntegralType`  (`ByteType`, `ShortType`, `IntegerType`, `LongType`) i.e. (`BYTE`, `SHORT`, `INT`, `LONG`) following the SQL standard by flooring `TIME` values to seconds since midnight.

  #### Changes:
  - Add `TIME` to integral casting logic in `Cast.scala` with proper overflow handling
  - Can potentially overflow on both `Short` and `Byte`
  - Add unit test coverage for `TIME` casting scenarios for both ansi and non-ansi
  - Update SQL integration tests with TIME casting examples for valid and error cases
  - Handle fractional seconds with "proper rounding" behavior - which is to truncate
  
  
     
     **Note:** `TimeType` to `IntegralType` largely follows `TimestampType` to integral casting for a few reasons:
  1. Both types are stored as `Long` internally (`TIME`: nanoseconds since midnight, `TIMESTAMP`: microseconds since epoch)
  2. Since they are both already stored as Long, there is a conversion to `Long` seconds as an intermediate step before casting to smaller integral types
  3. Uses similar overflow checking (i.e. `longValue == longValue.toByte`) to validate conversion safety
 
### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
  Currently, Spark SQL's`TIME` data type cannot be cast to integral types. The reasons for this functionality are:
  - Better overall SQL standard compliance
  - Interoperability/Code Migrations with other database systems that support `TIME` to integral casting
  - Also to enable simple arithmetic operations by using `TIME` values as integers

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as new features, bug fixes, or other behavior changes. Documentation-only updates are not considered user-facing changes.

If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes. This PR adds new casting functionality from `TimeType` -> (`Byte`, `Short`, `Int`, `Long`) that previously did not exist.  

 **Before this change:**
```scala
  spark.sql("SELECT CAST(TIME '10:10:10' AS INT)").show()
 /*
 org.apache.spark.sql.catalyst.ExtendedAnalysisException: [DATATYPE_MISMATCH.CAST_WITHOUT_SUGGESTION] Cannot resolve "CAST(TIME '10:10:10' AS INT)" due to data type mismatch: cannot cast "TIME(6)" to "INT". SQLSTATE: 42K09; line 1 pos 7;
'Project [unresolvedalias(cast(10:10:10 as int))]
*/
```
  After this change:
```sql
  SELECT CAST(TIME '10:10:10' AS INT)
  /**** Returns: 36610 (seconds since midnight) ****/

  SELECT CAST(TIME '00:00:04' AS TINYINT)
  /**** Returns: 4 ****/

  SELECT CAST(TIME '23:59:59' AS TINYINT)
  /**** Error: CAST_OVERFLOW - SQLSTATE: 22003 - when ansi flag is enabled - (86399 > 127) ****/
  /**** Returns: NULL when ansi flag is set to False - (86399 > 127) ****/

  Fractional seconds are truncated:
  SELECT CAST(TIME '00:00:01.7' AS INT)
  /**** Returns: 1 (floored not rounded up so 1 and not 2) ****/
  ```
### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
  1. Unit tests: Added comprehensive test coverage in CastSuiteBase, CastWithAnsiOnSuite, and CastWithAnsiOffSuite covering:
    - Basic `TimeType` to `IntegralType` casting
    - Overflow scenarios for smaller types (`ByteType`, `ShortType`)
    - Fractional second truncation
    - `ANSI` mode vs `non-ANSI` mode behavior
  
<img width="1111" height="572" alt="Screenshot 2025-07-11 at 11 15 37 PM" src="https://github.com/user-attachments/assets/ff258b22-8b33-4b4a-9a87-4be1855d5a50" />
<img width="1111" height="572" alt="Screenshot 2025-07-11 at 11 14 28 PM" src="https://github.com/user-attachments/assets/39ab5724-c868-4a67-889e-b064b242c52b" />

  2. SQL integration tests: Updated cast.sql with `TIME` casting examples and expected outputs - updated Golden Files and then ran test successfully
  
  ```bash
   SPARK_GENERATE_GOLDEN_FILES=1 ./build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z cast.sql"
 ./build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z cast.sql"
 ```
  3. Manual testing in Spark shell:
  ```scala
  val timeData = Seq(
    LocalTime.of(0, 0, 0),      // midnight = 0 seconds
    LocalTime.of(0, 0, 8),      // 8 seconds
    LocalTime.of(4, 41, 17),    // 16,877 seconds
    LocalTime.of(23, 59, 59)    // 86,399 seconds
  ).toDF("time_col")

  timeData.select(
    $"time_col",
    $"time_col".cast(IntegerType).alias("time_to_int"),
    $"time_col".cast(ByteType).alias("time_to_byte")
  ).show()

  // Results:
  // +--------+-----------+------------+
  // |time_col|time_to_int|time_to_byte|
  // +--------+-----------+------------+
  // |00:00:00|          0|           0|
  // |00:00:08|          8|           8|
  // |04:41:17|      16877|        NULL| // Overflow for BYTE
  // |23:59:59|      86399|        NULL| // Overflow for BYTE
  // +--------+-----------+------------+
```
  4. Fractional second testing manually in spark-shell:
  ```scala
  val edgeCases = Seq(
    LocalTime.of(0, 0, 0, 900000000),  // 0.9 seconds -> 0
    LocalTime.of(0, 0, 1, 700000000)   // 1.7 seconds -> 1
  ).toDF("time_col")

  edgeCases.select($"time_col", $"time_col".cast(IntegerType).alias("as_int")).show()
  /*
  +----------+------+
   |  time_col|as_int|
  +----------+------+
   |00:00:00.9|     0|
   |00:00:00.3|     0|
   |00:00:01.7|     1|
  +----------+------+
  */

```
  5. Overflow error message when there is overflow - manual test in spark-shell
  ```scala
  timeData.select(                                                                                                                                                                     $"time_col",                                                                                                                                                                                                                                                                                           
$"time_col".cast(ByteType).alias("time_to_byte")                                                                                                                            
).show()

timeData.select(                                                                                                                                                                     $"time_col",
        $"time_col".cast(ByteType).alias("time_to_byte")
      ).show()
     
     /*
org.apache.spark.SparkArithmeticException: [CAST_OVERFLOW] The value TIME '04:41:17' of the type "TIME(6)" cannot be cast to "TINYINT" due to an overflow. Use `try_cast` to tolerate overflow and return NULL instead. SQLSTATE: 22003
  at org.apache.spark.sql.errors.QueryExecutionErrors$.castingCauseOverflowError(QueryExecutionErrors.scala:87)
*/
``` 

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

Commit:
[SPARK-52619][SQL] Cast TimeType to IntegralType

  Implement casting from TIME data type to integral types (BYTE, SHORT, INT, LONG)
  following the SQL standard by rounding TIME values to seconds since midnight.

  Changes:
  - Add TIME to integral casting logic in Cast.scala with proper overflow handling
  - Add comprehensive test coverage for TIME casting scenarios
  - Update SQL integration tests with TIME casting examples
  - Handle fractional seconds with proper rounding behavior

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:
```diff
@@ -1507,4 +1507,26 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
+    checkEvaluation(cast(midnight, LongType), 0L)
+
+    // Test fractional seconds rounding
+    val fractionalTime = Literal.create(LocalTime.of(0, 0, 17, 500000000), TimeType(6))
```
Could you add a few tests for different precision of TimeType, please.

MaxGekk commented:
@uros-db Could you review this PR since you are working on similar one: https://github.com/apache/spark/pull/51455

uros-db commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastWithAnsiOffSuite.scala:
```diff
@@ -907,4 +907,18 @@ class CastWithAnsiOffSuite extends CastSuiteBase {
     }
   }
+
+  test("cast time to integral types with overflow") {
```
This looks like a duplicate test. Please consider putting it in `CastSuiteBase` instead.

uros-db commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastWithAnsiOnSuite.scala:
```diff
@@ -797,4 +797,18 @@ class CastWithAnsiOnSuite extends CastSuiteBase with QueryErrorsBase {
     }
   }
+
+  test("cast TimeType to integral types with overflow") {
```
This looks like a duplicate test. Please consider putting it in `CastSuiteBase` instead.

uros-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:
```diff
@@ -370,6 +372,7 @@ object Cast extends QueryErrorsBase {
     case (_, _: StringType) => false
 
     case (TimestampType, ByteType | ShortType | IntegerType) => true
+    case (_: TimeType, ByteType | ShortType) => true
```
Just checking - do we only want to catch `ByteType` & `ShortType` here?

uros-db reviewed (approved):
I left a few comments (mostly nits), otherwise LGTM.

Also, some tests are failing in GHA, but it may not be related.

Commit:
Added the tests with diffrent TIME precision and also used the constant for LocalTime.MIDNIGHT
Also changed the Timestamp(_) case to _: Timestamp

fartzy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:
```diff
@@ -1960,6 +2027,7 @@ case class Cast(
     case DateType =>
       (c, evPrim, evNull) => code"$evNull = true;"
     case TimestampType => castTimestampToIntegralTypeCode(ctx, "int", from, IntegerType)
+    case TimeType(_) => castTimeToIntegralTypeCode(ctx, "int", from, IntegerType)
```
Thanks for that catch! I changed to the more general `_: TimeType`

fartzy commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:
```diff
@@ -1507,4 +1507,26 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
+    checkEvaluation(cast(smallTime, LongType), 112L)
+
+    // Test edge cases
+    val midnight = Literal.create(LocalTime.of(0, 0, 0), TimeType(6)) // 0 seconds
```
Good idea! I changed it

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:
```diff
@@ -1507,4 +1507,52 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
+    checkEvaluation(cast(midnight, LongType), 0L)
+
+    // Test fractional seconds rounding
+    val fractionalTime = Literal.create(LocalTime.of(0, 0, 17, 500000000), TimeType(6))
```
Could you add more checks for rounding:
- .000001
- .999999
- .6
- .4
- .555

fartzy replied:
I added the tests, thanks for the suggestion

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastWithAnsiOnSuite.scala:
```diff
@@ -797,4 +797,25 @@ class CastWithAnsiOnSuite extends CastSuiteBase with QueryErrorsBase {
+
+    // Short and Byte should overflow and throw ArithmeticException (ANSI mode)
+    // 86399 > Short.MaxValue (32767) and > Byte.MaxValue (127)
+    checkExceptionInExpression[SparkArithmeticException](
```
Could you use `checkErrorInExpression` instead of the legacy `checkExceptionInExpression`, please. Error messages are not stable, and tech writers can change them. The tests shouldn't depend on error message formats. Ideally only from error message parameters.

fartzy replied:
I actually wondered if it seemed brittle too but I wasn't sure. Thanks for telling me about the `checkErrorInExpression`

fartzy commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:
```diff
@@ -1507,4 +1507,52 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
+    checkEvaluation(cast(noon0, IntegerType), 43200)
+    checkEvaluation(cast(noon2, IntegerType), 43200)
+    checkEvaluation(cast(noon4, IntegerType), 43200)
+
```
I removed the empty line sorry about that!

fartzy commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuiteBase.scala:
```diff
@@ -1507,4 +1507,52 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
+
+    checkEvaluation(cast(noon0, IntegerType), 43200)
+    checkEvaluation(cast(noon2, IntegerType), 43200)
+    checkEvaluation(cast(noon4, IntegerType), 43200)
```
Good idea! I added some fractional seconds

Commit:
Added some more tests and also changed to checkErrorInExpression instead of the legacy checkExceptionInExpression.
Also I deleted some blank lines and combined sections with the similar tests per the formatting standards on this file.
Also added some extra fraction to some of the tests with higher precision

MaxGekk commented:
+1, LGTM. Merging to master.
Thank you, @fartzy and @uros-db for review.
